reactive/kotlinx-coroutines-rx2/src/RxScheduler.kt

```diff
@@ -4,16 +4,129 @@
 
 package kotlinx.coroutines.rx2
 
-import io.reactivex.Scheduler
+import io.reactivex.*
+import io.reactivex.disposables.*
+import io.reactivex.plugins.*
 import kotlinx.coroutines.*
-import java.util.concurrent.TimeUnit
-import kotlin.coroutines.CoroutineContext
+import kotlinx.coroutines.channels.*
+import java.util.concurrent.*
+import kotlin.coroutines.*
 
 /**
  * Converts an instance of [Scheduler] to an implementation of [CoroutineDispatcher]
  * and provides native support of [delay] and [withTimeout].
  */
-public fun Scheduler.asCoroutineDispatcher(): SchedulerCoroutineDispatcher = SchedulerCoroutineDispatcher(this)
+public fun Scheduler.asCoroutineDispatcher(): CoroutineDispatcher =
+    if (this is DispatcherScheduler) {
+        dispatcher
+    } else {
+        SchedulerCoroutineDispatcher(this)
+    }
+
+/**
+ * Converts an instance of [CoroutineDispatcher] to an implementation of [Scheduler].
+ */
+public fun CoroutineDispatcher.asScheduler(): Scheduler =
+    if (this is SchedulerCoroutineDispatcher) {
+        scheduler
+    } else {
+        DispatcherScheduler(this)
+    }
+
+private class DispatcherScheduler(internal val dispatcher: CoroutineDispatcher) : Scheduler() {
+
+    private val job = SupervisorJob()
+    private val scope = CoroutineScope(job + dispatcher)
+
+    override fun scheduleDirect(block: Runnable): Disposable {
+        if (scope.isActive) {
+            return scope.launch {
+                block.runWithRx()
+            }.asDisposable()
+        }
+
+        return Disposables.disposed()
+    }
+
+    override fun scheduleDirect(block: Runnable, delay: Long, unit: TimeUnit): Disposable {
+        if (delay <= 0) {
+            return scheduleDirect(block)
+        }
+
+        if (scope.isActive) {
+            return scope.launch {
+                delay(unit.toMillis(delay))
+                block.runWithRx()
+            }.asDisposable()
+        }
+
+        return Disposables.disposed()
+    }
+
+    private fun Runnable.runWithRx() = RxJavaPlugins.onSchedule(this).run()
+
+    override fun createWorker(): Worker =
+        DispatcherWorker(dispatcher, job)
+
+    override fun shutdown() {
+        scope.cancel()
+    }
+
+    private class DispatcherWorker(dispatcher: CoroutineDispatcher, parentJob: Job) : Worker() {
+
+        private val workerScope = CoroutineScope(SupervisorJob(parentJob) + dispatcher)
+        private val blockChannel = Channel<Job>(Channel.UNLIMITED)
+        private var queueProcessingJob: Job? = null
+
+        override fun isDisposed(): Boolean = !workerScope.isActive
+
+        override fun schedule(block: Runnable): Disposable {
+            startProcessingQueue()
+            if (workerScope.isActive) {
+                val job = workerScope.launch(start = CoroutineStart.LAZY) {
+                    block.run()
+                }
+                blockChannel.offer(job)
```

Comment:
If the channel has been closed, this will throw an exception, which i don't think should be thrown from this method directly - maybe reported to `RxPlugins.onError`?

Reply:
I expect the channel to be closed at the same time as ```workerScope``` is set to inactive. So it should not even reach here if the channel has been closed. If it does I think it's better to let the exception be thrown so we can immediately know and fix.